Migration of L1 tracker primitives from SLHCDEV

delaere opened the pull request:
This PR introduces the tracker L1 primitives for phase2.

This is the first time that that code appears in a mainstream release, but it was developed and tested in SLHC and SLHCDEV releases for a long time.

Here, we introduce the core functionality: the TTCluster and TTStubs dataformats and related algorithms, which therefore only affects DataFormats and L1Trigger. The corresponding code in SimTracker, Validation and DQM is still being adapted to CMSSW_8_1_X. 

For now, only the main algorithms are functional. Others are introduced here but commented out for now and will become effective in a future development. Still, it is important to have the classes there already for bookkeeping reasons.

This PR is the basis for the track trigger developments and studies needed for the tracker TDR.

@boudoul , @sudhaahuja FYI

delaere committed:
step 1: Dataformat and L1Trigger code

at this stage, there is an error when building the dataformat library.
In the L1Trigger code, some stub algos are commented out.
In the L1Trigger tests, much is to be done (to be rewritten most probably).

delaere committed:
Fix BuildFile and class_def. Temporary checksums.

delaere committed:
Merged L1primitivesMigration_step1 from repository delaere

delaere committed:
some cleaning of classes.h and classes_def.xml

delaere committed:
Merged L1primitivesMigration_step1 from repository delaere

delaere committed:
adapt to the change of case in the TrackerTopology methods

delaere committed:
fix the classes_def.xml

delaere committed:
Cleaning and proper configuration of TTClusterBuilder

delaere committed:
Fix the TTCluster and TTStub builders

Properly iterate over stacks and use the TrackerTopology

delaere committed:
First step in adapting the Stub reco algos to new numbering

At this stage, it works with the window2013 algo. To make the tab2013 algo working, we need to pass the topology as well to the TTStubAlgo.

delaere committed:
Reenable tab2013 stub algo.

This implied to add a member to store the topology to the TTStubAlgorithm, and to adapt accordingly all derived classes.
Comments and debug output were also removed.

It runs and produces stubs when enabled in the sequence.
We should see how to adapt the sequence (change removeTTI to keep the stubs).

delaere committed:
Merged L1primitivesMigration_step1 from repository delaere

delaere committed:
Small adaptation to port TrackTrigger devs to 81X

A couple of methods changed namÃe (case)and the TrackerGeometry
had to be changed: ModuleType has to be known outside the class.

delaere committed:
Merged L1primitivesMigration_step1 from repository delaere

Conflicts:
	Geometry/TrackerGeometryBuilder/interface/TrackerGeometry.h
	Geometry/TrackerGeometryBuilder/src/TrackerGeometry.cc

solved by adopting the existing version (duplicate change)

cmsbuild commented:
A new Pull Request was created by  @delaere (Christophe Delaere) for CMSSW_8_1_X.

It involves the following packages:

DataFormats/L1TrackTrigger
L1Trigger/TrackTrigger


@cmsbuild, @mulhearn, @davidlange6 can you please review it and eventually sign? Thanks.
@Martin-Grunewald this is something you requested to watch as well.
@slava77, @degano, @smuzaffar you are the release manager for this.

cms-bot commands are list here cms-sw/cmssw#13028

davidlange6 commented on DataFormats/L1TrackTrigger/interface/TTCluster.h:
```diff
@@ -0,0 +1,185 @@
+
+  private:
+    /// Data members
+    std::vector< T >  theHits;
```
@delaere - is the maximum size of these vectors small and known?

[given the size of the PR, expect things little by little]

delaere replied:
Unfortunately not. In general the size is small but it can get very large in some cases. The only formal maximum is the number of strips in the module.

davidlange6 commented on DataFormats/L1TrackTrigger/interface/TTStub.h:
```diff
@@ -0,0 +1,187 @@
+    /// Helper methods: findABC( ... )
+
+    /// Clusters composing the Stub
+    std::vector< edm::Ref< edmNew::DetSetVector< TTCluster< T > >, TTCluster< T > > > getClusterRefs() const { return theClusterRefs; }
```
what is this meant to physically represent?

delaere replied:
A stub is a pair of clusters with some additional information (offset, etc.).
I changed the implementation in an update of this PR, where the cluster refs are kept in a lighter way (no need for a vector).

davidlange6 commented on DataFormats/L1TrackTrigger/src/TTCluster.cc:
```diff
@@ -0,0 +1,120 @@
+    int row = 0;
+    if ( this->getRows().size() == 0 )
+    {
+      row = theHits.at(i)->row();
```
use theHits[i] please

davidlange6 commented on DataFormats/L1TrackTrigger/src/TTCluster.cc:
```diff
@@ -0,0 +1,120 @@
+  std::vector< int > temp;
+  for ( unsigned int i = 0; i < theHits.size(); i++ )
+  {
+    temp.push_back( theHits.at(i)->row() );
```
add temp.reserve(theHits.size())

davidlange6 commented on DataFormats/L1TrackTrigger/src/TTCluster.cc:
```diff
@@ -0,0 +1,120 @@
+  std::vector< int > temp;
+  for ( unsigned int i = 0; i < theHits.size(); i++ )
+  {
+    temp.push_back( theHits.at(i)->column() );
```
and same here.

davidlange6 commented on L1Trigger/TrackTrigger/interface/TTClusterAlgorithm.h:
```diff
@@ -0,0 +1,78 @@
+{
+  protected:
+    /// Data members
+    std::string                  className_;
```
could this be an enum?

delaere replied:
This is only used for printout... if this is a problem, it could be removed all together, I guess.

davidlange6 commented on L1Trigger/TrackTrigger/interface/TTClusterAlgorithm_2d.h:
```diff
@@ -0,0 +1,129 @@
+{
+  private:
+    /// Data members
+    boost::shared_ptr< TTClusterAlgorithm< T > > _theAlgo;
```
use std::shared_ptr

davidlange6 commented on L1Trigger/TrackTrigger/interface/TTClusterAlgorithm_a.h:
```diff
@@ -0,0 +1,119 @@
+/*! \class   TTClusterAlgorithm_a
```
could the algorithms get descriptive names? (not dates, letters)

davidlange6 commented on L1Trigger/TrackTrigger/interface/TTClusterAlgorithm_a.h:
```diff
@@ -0,0 +1,119 @@
+  {
+    std::vector< T > temp;
+    temp.push_back(*inputIterator);
+    output.push_back(temp);
```
output.reserve(input.size());

davidlange6 commented on L1Trigger/TrackTrigger/interface/TTStubAlgorithm_globalgeometry.h:
```diff
@@ -0,0 +1,131 @@
+      /// Get magnetic field
+      edm::ESHandle< MagneticField > magnet;
+      record.getRecord< IdealMagneticFieldRecord >().get(magnet);
+      double mMagneticFieldStrength = magnet->inTesla( GlobalPoint(0,0,0) ).z();
```
hum, there should be a better way to get this (but I don't know it)

delaere replied:
the code snipped above is used in many places in CMSSW, but I found the following alternative, that I think is equivalent:`double mMagneticFieldStrength = std::abs(magnet->nominalValue());`

davidlange6 commented on L1Trigger/TrackTrigger/interface/TTStubAlgorithm_window2012.h:
```diff
@@ -0,0 +1,135 @@
+    virtual ~ES_TTStubAlgorithm_window2012(){}
+
+    /// Implement the producer
+    boost::shared_ptr< TTStubAlgorithm< T > > produce( const TTStubAlgorithmRecord & record )
```
there is a lot of repeated code in produce Could it be put into the base class?

davidlange6 commented on L1Trigger/TrackTrigger/plugins/TTStubBuilder.h:
```diff
@@ -0,0 +1,393 @@
+
+  for (TrackerGeometry::DetContainer::const_iterator gd=theTrackerGeom->dets().begin(); gd != theTrackerGeom->dets().end(); gd++) {
+      DetId detid = (*gd)->geographicalId();
+      if(detid.subdetId()==1 || detid.subdetId()==2 ) continue; // only run on OT
```
pls use tracker enums instead of 1 and 2

davidlange6 commented on L1Trigger/TrackTrigger/plugins/TTStubBuilder.h:
```diff
@@ -0,0 +1,393 @@
+            {
+              /// No, so new entry
+              std::vector< TTStub< T > > tempStubs;
+              tempStubs.clear();
```
not needed.

davidlange6 commented on L1Trigger/TrackTrigger/plugins/TTStubBuilder.h:
```diff
@@ -0,0 +1,393 @@
+          std::vector< std::pair< unsigned int, double > > bendMap;
+          for ( unsigned int i = 0; i < is.second.size(); ++i )
+          {
+            bendMap.push_back( std::pair< unsigned int, double >( i, is.second[i].getTriggerBend() ) );
```
reserved(is.second.size())

davidlange6 commented:
for this to run without crashing, the consumes migration is needed.

delaere commented:
Hi David. Thanks for your comments. This is not my code but I will look to fix the code in the best possible way.

delaere committed:
Review the implementation of TTCluster and TTStub

Addressing comments from David.
Among other things, changed the way clusters are stored in the stub. No need for a vector.

cmsbuild commented:
Pull request #13696 was updated. @cmsbuild, @rekovic, @mulhearn, @davidlange6 can you please check and sign again.

delaere commented:
I modified the DataFormat part... still have to work on the algo part.

delaere committed:
Work on Cluster and Stub builders. More to come.

- consumes migration
- auto for iterators
- other comments from David

This only touches the plugin code, not the actual algorithms.
I contacted the L1 tracking group regarding that... waiting for an answer.

The main points to be addressed are:
1. Do we need all algos?
2. Can the name be changed to something more informative?

cmsbuild commented:
Pull request #13696 was updated. @cmsbuild, @rekovic, @mulhearn, @davidlange6 can you please check and sign again.

delaere committed:
replace boost::shared_ptr by the std version

delaere committed:
Comments from David...

cmsbuild commented:
Pull request #13696 was updated. @cmsbuild, @rekovic, @mulhearn, @davidlange6 can you please check and sign again.

davidlange6 commented:
please test

cmsbuild commented:
The tests are being triggered in jenkins.

cmsbuild commented:
+1
Tested at: 7d078bff100f7ba4232e16e928755036ec16582e
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-13696/12251/summary.html

cmsbuild commented:
Comparison is ready
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-13696/12251/summary.html

delaere committed:
Cleaning of unused algorithms

Following a discussion with the L1TT group on hypernews, we converged on a minimal set to be kept in CMSSW_8_1_X.
https://hypernews.cern.ch/HyperNews/CMS/get/l1-track-trigger/82/2.html

TTClusterAlgorithm_2d2013.h     --> KEEP: TTClusterAlgorithm_official.h
TTClusterAlgorithm_2d.h         --> REMOVE
TTClusterAlgorithm_a.h          --> REMOVE
TTClusterAlgorithm_broadside.h  --> REMOVE
TTClusterAlgorithm_neighbor.h   --> KEEP

TTStubAlgorithm_a.h              --> REMOVE
TTStubAlgorithm_cbc3.h           --> KEEP
TTStubAlgorithm_globalgeometry.h --> REMOVE
TTStubAlgorithm_pixelray.h       --> REMOVE
TTStubAlgorithm_tab2013.h        --> KEEP: TTStubAlgorithm_official.h
TTStubAlgorithm_window2012.h     --> REMOVE
TTStubAlgorithm_window2013.h     --> REMOVE
TTStubAlgorithm_window.h         --> REMOVE

cmsbuild commented:
Pull request #13696 was updated. @cmsbuild, @rekovic, @mulhearn, @davidlange6 can you please check and sign again.

delaere commented:
Dear all,

With this last commit, the PR should be ok to proceed, now. All comments raised so far have been addressed or answered above.

davidlange6 commented:
please test

cmsbuild commented:
The tests are being triggered in jenkins.

cmsbuild commented:
+1
Tested at: e0678153bc629e272dc12c22e91953618dc2bbc7
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-13696/12499/summary.html

cmsbuild commented:
Comparison is ready
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-13696/12499/summary.html

delaere commented:
Dear all,

Could somebody have a look and let me know if there are remaining issues?
Thanks!
